### ISSUE TITLE
Use jupyverse as a "real" server

### ISSUE THREAD

**Issue by davidbrochart** [jupyterlite/jupyterlite#779]
### Problem

AFAIU, JupyterLite doesn't use any of the Jupyter server infrastructure. This means that a lot of things are and will be re-implemented, such as authentication/authorization.

### Proposed Solution

[Jupyverse](https://github.com/jupyter-server/jupyverse) is an ASGI Jupyter server. It can use any ASGI web server under the hood, such as [uvicorn](https://github.com/encode/uvicorn). It could be interesting if JupyterLite could communicate with jupyverse directly at the ASGI level, through a custom ASGI server which would simply make connections and pass events.
Maybe even the kernel protocol could be transmitted directly over an ASGI extension, instead of being serialized/deserialized and transmitted over WebSocket.

**Comment by bollwyvl** [comment]
> doesn't use any of the Jupyter server infrastructure

Indeed: jupyterlite doesn't use _any_ "real server" infrastructure beyond HTTPS hosting. `jupyter lite serve` is only meant for local testing, to help avoid the insecure-by-default behavior of `python -m http.server`.

Not having a bespoke server process is one of the **primary features** of jupyterlite. It's a far easier task to stand up a jupyterlite site on a free host with transparently configured CDN (readthedocs, git(hub|lab) pages) that can easily serve thousands of "simultaneous" users vs any python daemon-backed system.

Adding some other backend for storage presently would require replacing the existing localftorage+pre-cached content, but #315 seeks to improve that to allow for a site deployer to compose various strategies.

> things are and will be re-implemented

The price we pay for re-implementing the parts that make any sense at all is that the total weight of the "server" is in the tens-to-hundreds of kilobytes, which gets cached very robustly in a serviceworker. Standing up an environment in the browser is _expensive_ and we're presently trying to move _away_ from any language runtimes in the core repository, aside from _perhaps_ JS.

If anything, refactoring `jupyter-contents` into a well-typed, standalone, asynchronous component that could be used by both `jupyter_server` and `jupyverse` at runtime, and `jupyter lite build` (not at runtime) would be the only path of particular interest.

> JupyterLite could communicate with jupyverse directly

I don't really know what this would do, but it can almost certainly be done as an extension: we're unlikely to add _any_ specific server integration to this repo.

**Comment by davidbrochart** [comment]
I guess I  didn't make it clear that jupyverse wouldn't be used as a web server, such as one running in a process. Only the application part would be used, and it would all run in the browser.
I agree that it would make JupyterLite depend on a Python runtime, such as Pyodide, so it would be preferable to have it as an extension. In the case JupyterLite ships a Python kernel, this would basically come for free though.

**Comment by bollwyvl** [comment]
> run in the browser.

Too many things related to sockets, zmq, etc. are just not going to "work" for downstream, third-party extensions at this point, no matter how carefully they are architected. So whatever we had to do to make the entire starlette stack + dog work would probably not look much like what's running on top of the API for other users.

This is partially academic at this point, of course, as i don't think there _are_ any third-party extensions for jupyverse. If this was worth doing, it would be worth doing against `jupyter_server` so that the hundreds of existing extensions.

> for free  

Except for the 120mb of RAM that an extra worker takes up, the 5mb for a cold boot of pyodide, and then howver many tens of megabytes the whole `starlette` + `fastapi` + whatever, all before drawing the first effective pixels of interactive computing (e.g. listing contents, kernelspecs).

> preferable to have it as an extension

Actually, the `server` application is presently not replaceable, so it would have to be a hard fork at this point, as I really can't see this making sense in core at this time. In a year or two, it could be feasible to consider... 

However, in a year or two, it could be feasible to build a modular, well-typed, multi-language WASM-delivered server/kernel stack for the browser which could _also_ run on the server/cluster/serverless, which is where I actually think this technology will eventually go.

**Comment by jtpio** [comment]
A couple of months ago I was also optimistic about code reuse and trying to leverage (parts of) `jupyverse` for the in-browser server: https://github.com/jupyterlite/jupyterlite/discussions/372

However as Nick said above there are many more advantages of keeping the custom (still lightweight) at least for now. It's easier to debug and starts off much quicker that a Pyodide or Xeus Python based Python running the browser.
